bmcd/src/legacy.rs

```diff
@@ -0,0 +1,395 @@
+    let node = match get_node_param(&query) {
+        Ok(n) => n,
+        Err(e) => return HttpResponse::BadRequest().body(e),
+    };
+
+    let (tx, mut rx) = mpsc::channel(64);
+
+    let bmc = bmc.clone();
+    let handle = rt::spawn(async move { bmc.set_node_in_msd(node, UsbRoute::Bmc, tx).await });
+
+    let logger = rt::spawn(async move {
+        while let Some(msg) = rx.recv().await {
+            log::info!("{}", msg);
+        }
+    });
+
+    tokio::try_join!(handle, logger).to_response("await threads")
+}
+
+fn get_node_param(query: &Query) -> Result<NodeId, &'static str> {
+    let Some(node_str) = query.get("node") else {
+        return Err("Missing `node` parameter");
+    };
+
+    let Ok(node_num) = i32::from_str(node_str) else {
+        return Err("Parameter `node` is not a number");
+    };
+
+    let Ok(node) = node_num.try_into() else {
+        return Err("Parameter `node` is out of range 0..3 of node IDs");
+    };
+
+    Ok(node)
+}
+
+async fn get_system_information() -> HttpResponse {
+    let version = env!("CARGO_PKG_VERSION");
+    let build_time = build_time::build_time_utc!("%Y-%m-%d %H:%M:%S-00:00");
+    let ipv4 = get_ipv4_address().unwrap_or("Unknown".to_owned());
+    let mac = get_mac_address().await;
+
+    let body = json!({
+        "response": [{
+            "version": version,
+            "buildtime": build_time,
+            "ip": ipv4,
+            "mac": mac,
+        }]
+    });
+
+    HttpResponse::Ok().json(body)
+}
+
+fn get_ipv4_address() -> Option<String> {
+    for interface in if_addrs::get_if_addrs().ok()? {
+        // TODO: for compatibility reasons, only IPv4 of eth0 is returned. Ideally, both IPv4 and
+        // IPv6 addresses of all non-loopback interfaces should be returned.
+        if interface.is_loopback() || interface.name != "eth0" {
+            continue;
+        }
+
+        let std::net::IpAddr::V4(ip) = interface.ip() else {
+            continue;
+        };
+
+        return Some(format!("{}", ip));
+    }
+
+    None
+}
+
+async fn get_mac_address() -> String {
+    tokio::fs::read_to_string("/sys/class/net/eth0/address")
+        .await
+        .unwrap_or("Unknown".to_owned())
+}
+
+async fn set_node_power(bmc: &mut Arc<BmcApplication>, query: Query) -> HttpResponse {
+    for id in 0..4 {
+        let param = format!("node{}", id);
+        let req_status = match query.get(&param).cloned().as_deref() {
+            Some("0") => false,
+            Some("1") => true,
+            Some(x) => {
+                let msg = format!("Invalid value `{}` for parameter `{}`", x, param);
+                return HttpResponse::BadRequest().body(msg);
+            }
+            None => continue,
+        };
+        // Unwrap: values of the loop are statically bound to cover only allowed NodeId values
+        let node = NodeId::try_from(id).unwrap();
+        let curr_status = match bmc.get_node_power(node).await {
```

Comment:
The logic of "checking if a node is already in the state you want it to transition to", is redundant and can be removed. It's already done by the `PowerController.` I would like to keep things as straightforward as possible in this routine. The only function it "should" have is to parse from an HTTP query to a corresponding method.